ironfish/src/wallet/scanner/walletScanner.ts

```diff
@@ -137,71 +167,43 @@ export class WalletScanner {
       )
     }
 
-    const accounts = await AsyncUtils.filter(this.wallet.listAccounts(), async (account) => {
-      if (!account.scanningEnabled) {
-        return false
-      }
-
-      const accountHead = await account.getHead()
-
-      if (!accountHead) {
-        return blockHeader.sequence === 1
-      } else {
-        return BufferUtils.equalsNullable(accountHead.hash, blockHeader.previousBlockHash)
-      }
-    })
-
-    const shouldDecryptAccounts = accounts.filter((a) =>
-      this.wallet.shouldDecryptForAccount(blockHeader, a),
-    )
-
-    const shouldDecryptAccountIds = new Set(shouldDecryptAccounts.map((a) => a.id))
-
-    const decryptedTransactions = await Promise.all(
-      getTransactionsWithNoteIndex(blockHeader, transactions).map(
-        ({ transaction, initialNoteIndex }) =>
-          this.wallet
-            .decryptNotes(transaction, initialNoteIndex, false, shouldDecryptAccounts)
-            .then((r) => ({
-              result: r,
-              transaction,
-            })),
-      ),
-    )
-
-    // account id -> transaction hash -> Array<DecryptedNote>
-    const decryptedNotesMap: Map<string, BufferMap<Array<DecryptedNote>>> = new Map()
-    for (const { transaction, result } of decryptedTransactions) {
-      for (const [accountId, decryptedNotes] of result) {
-        const accountTxnsMap =
-          decryptedNotesMap.get(accountId) ?? new BufferMap<Array<DecryptedNote>>()
-        accountTxnsMap.set(transaction.hash(), decryptedNotes)
-        decryptedNotesMap.set(accountId, accountTxnsMap)
+    const accounts = new Array<Account>()
+    for (const candidate of this.scanningAccounts) {
+      if (
+        !candidate.scanFrom ||
+        BufferUtils.equalsNullable(candidate.scanFrom.hash, blockHeader.previousBlockHash)
+      ) {
+        candidate.scanFrom = null
+        accounts.push(candidate.account)
       }
     }
 
-    for (const account of accounts) {
-      if (abort?.signal.aborted) {
-        return
-      }
+    return decryptor.decryptNotesFromBlock(
+      blockHeader,
+      transactions,
+      accounts,
+      async (decryptedNotesMap) => {
+        for (const account of accounts) {
+          if (abort?.signal.aborted) {
+            return
+          }
 
-      const accountTxnsMap = decryptedNotesMap.get(account.id)
+          const accountTxnsMap = decryptedNotesMap.get(account.id)
 
-      const txns = transactions.map((transaction) => ({
-        transaction,
-        decryptedNotes: accountTxnsMap?.get(transaction.hash()) ?? [],
-      }))
+          const txns = transactions.map((transaction) => ({
+            transaction,
+            decryptedNotes: accountTxnsMap?.get(transaction.hash()) ?? [],
+          }))
 
-      await this.wallet.connectBlockForAccount(
-        account,
-        blockHeader,
-        txns,
-        shouldDecryptAccountIds.has(account.id),
-      )
-    }
+          const shouldDecrypt = this.wallet.shouldDecryptForAccount(blockHeader, account)
```

Comment:
Do we check `shouldDecryptForAccount` before doing decryption? We compare the `account.createdAt` to the current block to avoid doing any decryption if the block was added to the chain before the account's `createdAt`

Reply:
Good point, not sure how I missed this. I wonder if it's even possible for an account to have their head set before their `createdAt`? That would allow simplifying the logic a bit

Reply:
The head can be set after the birthday if you set it to the birthday, then immediately rewind because you were on a fork.

Reply:
I understand the fork case, but I don't think that checking for the sequence (like `shouldDecryptForAccount` is doing) is the right approach. Consider this chain:

```mermaid
graph LR;
  a1["A1 (seq: 100)"]
  a1-->b1["B1 (seq: 101)"]
  b1-->b2["B2 (seq: 102)"]
  b2-->b3["B3 (seq: 103)"]
  b3-->b4["B4 (seq: 104)"]
  a1-->c1["C1 (seq: 101)"]
  c1-->c2["C2 (seq: 102)"]
  c2-->c3["C3 (seq: 103)"]
```

Here we have a chain A that forks and is split into two chains: B and C. Suppose an account is initially created at block B3, therefore at sequence 103. The main chain then becomes C, the blocks are rolled back and a transaction involving the account is created at block C2, sequence 102.

In this case, the `createdAt` has sequence 103, but a transaction exists at sequence 102. `shouldDecryptForAccount` will return `false` for C2, and will cause the scanner to miss the transaction.

Anyway, this problem is not strictly relevant to this PR, but if the intent of `shouldDecryptForAccount` is to handle forks, then we probably need to improve on that.

Reply:
The documented above behaviour is expected. It's why the birthday is not actually 103 for B3.. The createdAt sequence should correctly be 103 - confirmation range which is defaulted to 2. The birthday should be calculated at 101. That being said, it's still possible to miss if at the point of creation you are on a re-org that's longer than the confirmation time. However, the idea of the confirmation time is to make the unlikely. This is a change that Hugh is working on.

The other note is that in general, birthdays are heuristics and if they are wrong you are expected to miss transactions.